src/ast/nodes/ImportExpression.ts

```diff
@@ -45,6 +54,95 @@ export default class ImportExpression extends NodeBase {
 		this.source.bind();
 	}
 
+	/**
+	 * Get imported variables for deterministic usage, valid cases are:
+	 *
+	 * - `const { foo } = await import('bar')`.
+	 * - `(await import('bar')).foo`
+	 * - `import('bar').then(({ foo }) => {})`
+	 *
+	 * Returns empty array if it's side-effect only import.
+	 * Returns undefined if it's not fully deterministic.
+	 */
+	getDeterministicImportedNames(): readonly string[] | undefined {
+		// side-effect only
+		if (this.parent.type === 'ExpressionStatement') {
```

Comment:
This looks like a three-case switch-statement over `this.parent.type`, which would also avoid the repeated property accesses.

Alternatively, a better approach that is also used in other places and which solves TypeScript issues would be to use instanceof:

```ts
const {parent} = this;
if (parent instanceof ExpressionStatement) {
  return EMPTY_ARRAY;
}
if (parent instanceof AwaitExpression) {
  // Now the type is narrowed
  // It is always a good idea to avoid repeated property accesses for both performance and typing reasons
  const awaitParent = parent.parent;
  if (awaitParent instanceof ExpressionStatement) {
    return EMPTY_ARRAY;
  }
  if (awaitParent instanceof VariableDeclarator) {
    // Here the type-narrowing means we have id correctly typed
    if (!(awaitParent.id instanceof ObjectPattern)) return;
    ...
  }
}
```

One thing to look out for is if this creates cyclic dependencies. In that case, you need to check `type` after all. But otherwise, it should be marginally more performant and always has the right types.